### DISCUSSION TITLE
Counting in a for-loop | useful for trial numbers

### DISCUSSION THREAD

**Discussion by Adsecularis** [ctreffe/alfred#170] [Show and tell]
Hey there, I wanted to show off this little thing.
What it does is add 1 to the variable x each time it finishes the loop. That's it. It's a basic ability in Python so I guess you won't find something like it in the Alfred3 documentation. However it can be really important in an experiment.

For example we wanted to show participants 9 different texts in a random order, so we made a list with 9 integers and shuffled that randomly at the beginning of each experiment. But then what variable do I use for the trial? I can't use the shuffled list because reading text 5, then text 3 and then text 9 makes no sense to the participants. Also the list-integer doesn't tell me in which trial participants read which text unless I look at the whole list. I could solve that by making an identical list, which I do not shuffle to show me and the participants the correct trial number. But then I always have to manage shuffled lists and their non-shuffled counterparts.

So instead I use x, then the loop counts for me which trial it is at.
```python
import random
import alfred3 as al
exp = al.Experiment()

list = list(range(9))
random.shuffle(list)

@exp.member
class DemoPage(al.Page):
    def on_exp_access(self):
        x = 1
        for i in list:
            self += al.Text(f"I am the number in the list: {i}  |  I am the number of the trial: {x}")
            x += 1
        return x

if __name__ == "__main__":
    exp.run()
```

The same can be done in the for-loop for a section where i and x can be added to the vargs to be used in the pages of that loop like so:
```python
import random
import alfred3 as al
exp = al.Experiment()

list = list(range(9))
random.shuffle(list)

class TextPage(al.Page):
    def on_first_show(self):
        self += al.Text(f"I am text page number {self.vargs.x}, but I am actually showing text number {self.vargs.i+1}")

@exp.member
class DemoSection(al.Section):

    def on_exp_access(self):
        x = 1
        for i in list:
            self += TextPage(f"Text {x} of 9", name=f"text_{i+1}", vargs={"i": i, "x": x})
            x += 1
        return x

if __name__ == "__main__":
    exp.run()
```

**Comment by jobrachem** [comment]
Hi @Adsecularis, thanks so much for sharing your code! 😊 Please keep doing this! I think your code looks good. It shows that you are creatively employing Python's flexibility to solve problems in an elegant way. And on top of that, sharing your solutions is such a nice thing to do! 

I would like to share a few comments to build on your input.

1. It's quite risky to re-assign the names of builtin functions like `list`, as you do in your code examples. **Do not do this**. This can lead to confusion later on in your script, because you may want to use the actual function - but now it has been reassigned to a specific list. Such situations can be very hard to debug. So I would recommend that you choose a more concrete name like `trial_indices`, or something like that. Most IDEs like PyCharm have special highlighting for builtin functions, but there's also a [list](https://docs.python.org/3/library/functions.html).
2. If you are randomizing trials, you should always take care that you save both, the *position* (your `x`) and the actual identifying number (your `i`) in some way. I guess you are doing that anyway, but I want to point this fact out for other people that may be reading this thread. You can find guidance on how to implement this [here in to docs](https://alfredo3.psych.bio.uni-goettingen.de/docs/howto_loops.html#shuffle-the-order-of-pages)
3. Python has a nice builtin function called `enumerate`, that you can use to make your code even more efficient 😊 See below. (More [here at realpython.com](https://realpython.com/python-enumerate/)) 
4. Your `on_exp_access` method returns the counter `x` - that is not necessary, since the return value of this hook will not be used by any function within alfred3.
5. It is a good, precautionary practice to use only few, clearly constant objects as global objects. That is (similar to point 1) a precaution against confusion (or unforeseen effects). For this reason, I would recommend moving the definition and shuffling of the trial_number list (your object `list`) to a `Experiment.setup` hook and saving the list in the `Experiment.adata` dictionary, as in the code example below. This has the additional benefit that the shuffled list will be automatically saved to each session's dataset.

Example for `enumerate` ([docs](https://docs.python.org/3/library/functions.html#enumerate)) :
```python
values = ["a", "b", "c"]
>>> for count, value in enumerate(values):
...     print(count, value)
...
0 a
1 b
2 c
```

Updated code example:

```python
import random
import alfred3 as al
exp = al.Experiment()

@exp.setup
def setup(exp):
    trial_indices = list(range(9))
    random.shuffle(trial_indices)
    exp.adata["trial_indices"] = trial_indices

@exp.member
class DemoPage(al.Page):
    def on_exp_access(self):
        trial_indices = self.exp.adata["trial_indices"]

        for counter, trial_number in enumerate(trial_indices, start=1):
            self += al.Text(f"I am the number in the list: {trial_number}  |  I am the number of the trial: {counter}")

if __name__ == "__main__":
    exp.run()
```

**Reply by Adsecularis** [reply]
Thank you very much! 
My work on the current experiment is already done, but I will utilize your advice in the next.
